controller_manager_tests/src/extensible_controllers.cpp

```diff
@@ -0,0 +1,59 @@
+
+#include <controller_manager_tests/extensible_controllers.h>
+
+using namespace controller_manager_tests;
+
+bool ExtensibleController::init(hardware_interface::RobotHW* robot_hw,
+          ros::NodeHandle& root_nh, ros::NodeHandle& controller_nh)
+{
+  std::string vel_joint_name;
+  controller_nh.getParam("velocity_joint", vel_joint_name);
+
+  hardware_interface::VelocityJointInterface* joint_iface = robot_hw->get<hardware_interface::VelocityJointInterface>();
+  joint_iface->getHandle(vel_joint_name);
+  return true;
+}
+
+int ExtensibleController::helper()
+{
+  return 1;
+}
+
+void ExtensibleController::update(const ros::Time&, const ros::Duration&)
+{
+  // TODO: Publish this return value so that the test can validate it?
+  int foo = helper();
+}
+
+
+bool DerivedController::initRequest(hardware_interface::RobotHW* hw, ros::NodeHandle& nh, ros::NodeHandle& pnh,
+    controller_interface::ControllerBase::ClaimedResources& cr)
+{
+  return DerivedControllerInterface::initRequest(hw, nh, pnh, cr);
+}
```

Comment:
Having to pass through this call is my least favourite aspect of the whole business, but I can't really see a good way to get around it. It's not sufficient to just `using DerivedControllerInterface::initRequest;`, since that doesn't count as an actual override, so the main other alternative would be introducing some new template that would supply this function and take the things you want to extend as its arguments. I'm just not sure the savings in code is worth further obscuring what's really going on in here.

Reply:
(Of course, this approach of calling the other `init` method from `DerivedController::init` only works because `DerivedControllerInterface` is an intentional superset of the original controller's interface. If that were not the case, potentially this method would have to call multiple `initRequest` methods, another argument against trying to overly standardize it...)

Reply:
Follow up: It actually doesn't work to call both `initRequest` methods, since `init` is virtual, so they end up both calling the final class's one, rather than ones which correspond to their own classes. Given this, it's possible that indeed a universal `ExtensionInterfaceController` template (like `MultiInterfaceController`) could be developed, where you list only your new interface and all controllers you're extending, and it supplies an `initRequest` implementation which takes care of calling in order the init methods for the controllers you're extending.

Anyway, all of that would be API sugar on the basic principles already present in this PR.

Reply:
`ros_control` already relies heavily on inheritance, I think it's better for this case than composition.

Reply:
For any future readers, PR #301 is a general-purpose `CompositeController` which allows combining multiple controllers into one, and cleanly handles details like calling the subcontroller start and stop methods in order, etc.

Regarding the point made above about `init` being virtual— it actually _does_ work out to call the individual subcontrollers' `initRequest` methods, because `init` is a virtual member of the templated `Controller<>`/`MultiInterfaceController<>` classes, _not_ of `ControllerBase`. So if only `ControllerBase` is inherited with `public virtual`, then there will still be multiple instances of all the other classes _other than `ControllerBase`, which means exactly one initRequest/init function per subcontroller. This reality is what made the `CompositeController` template possible.